src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/AppleCertificatePal.Pkcs12.cs

```diff
@@ -0,0 +1,140 @@
+using Microsoft.Win32.SafeHandles;
+
+namespace Internal.Cryptography.Pal
+{
+    internal sealed partial class AppleCertificatePal : ICertificatePal
+    {
+        private static ICertificatePal ImportPkcs12(
+            byte[] rawData,
+            SafePasswordHandle password,
+            bool exportable,
+            SafeKeychainHandle keychain)
+        {
+            using (ApplePkcs12Reader reader = new ApplePkcs12Reader(rawData))
+            {
+                reader.Decrypt(password);
+
+                UnixPkcs12Reader.CertAndKey certAndKey = reader.GetSingleCert();
+                AppleCertificatePal pal = (AppleCertificatePal)certAndKey.Cert;
+
+                SafeSecKeyRefHandle safeSecKeyRefHandle =
+                    ApplePkcs12Reader.GetPrivateKey(certAndKey.Key);
+
+                AppleCertificatePal newPal;
+
+                using (safeSecKeyRefHandle)
+                {
+                    // SecItemImport doesn't seem to respect non-exportable import for PKCS#8,
+                    // only PKCS#12.
+                    //
+                    // So, as part of reading this PKCS#12 we now need to write the minimum
+                    // PKCS#12 in a normalized form, and ask the OS to import it.
+                    if (!exportable && safeSecKeyRefHandle != null)
+                    {
+                        return ImportPkcs12NonExportable(pal, safeSecKeyRefHandle, password, keychain);
+                    }
+
+                    newPal = pal.MoveToKeychain(keychain, safeSecKeyRefHandle);
+                }
+
+                // If no new PAL came back, it means we moved the cert, but had no private key.
+                return newPal ?? pal;
+            }
+        }
+
+        internal static ICertificatePal ImportPkcs12NonExportable(
+            AppleCertificatePal cert,
+            SafeSecKeyRefHandle privateKey,
+            SafePasswordHandle password,
+            SafeKeychainHandle keychain)
+        {
+            Pkcs12SmallExport exporter = new Pkcs12SmallExport(new TempExportPal(cert), privateKey);
+            byte[] smallPfx = exporter.Export(X509ContentType.Pkcs12, password);
+
+            SafeSecIdentityHandle identityHandle;
+            SafeSecCertificateHandle certHandle = Interop.AppleCrypto.X509ImportCertificate(
+                smallPfx,
+                X509ContentType.Pkcs12,
+                password,
+                keychain,
+                exportable: false,
+                out identityHandle);
+
+            // On Windows and Linux if a PFX uses a LocalKeyId to bind the wrong key to a cert, the
+            // nonsensical object of "this cert, that key" is returned.
+            //
+            // On macOS, because we can't forge CFIdentityRefs without the keychain, we're subject to
+            // Apple's more stringent matching of a consistent keypair.
+            if (identityHandle.IsInvalid)
+            {
+                identityHandle.Dispose();
```

Comment:
why do we dispose invalid handle?

Reply:
The P/Invoke marshaller always returns non-null SafeHandle values.  The invalid ones are still finalizable, unless you call Dispose or GC.SuppressFinalize.  Not doing so creates finalization pressure.